### PR TITLE
feat: Feat/9492 add unsupported method and legacy method middlewares

### LINKED ISSUE #9492: Do not implement RPC methods ahead of the permission middleware
Ref: https://github.com/MetaMask/metamask-extension/issues/24331

In the extension, some RPC methods, e.g. `wallet_switchEthereumChain` and `wallet_addEthereumChain`, are implemented ahead of the permission middleware. This makes it difficult to keep track of our restricted vs. unrestricted methods, while also literally circumventing our permission system. We are going to stop doing this in the extension (see above issue), and, once mobile has implemented the permission middleware, we shouldn't do it here either. Instead, we should reorder our `json-rpc-engine` middleware stack such that every method hits the permission middleware, and all unrestricted methods have to be enumerated in an unrestricted methods list.

### PR DESCRIPTION
**Pull request by Cal-L**
## **Description**

This PR is a carries over the changes from extension - https://github.com/MetaMask/metamask-extension/pull/24472 where new new middlewares are added and permission middleware reordered within the JSON rpc engine stack. The following changes are:
- New middleware for handling unsupported RPC methods. The only unsupported method at the moment is `eth_signTransaction`
- New middleware for handling legacy RPC methods. The only legacy RPC method where we need to preserve legacy logic is `eth_accounts`
- Moved permission middleware ahead of RPCMethodMiddleware aka user/wallet facing middleware

## **Related issues**

Fixes: #9492 

## **Manual testing steps**

For testing unsupported method middleware, you can update `UNSUPPORTED_RPC_METHODS` to include a method such as `wallet_getPermissions`
- Go on the MM test dapp
- Tap on GET_PERMISSIONS button
- Notice the error `Method not supported`

For testing legacy method middleware, test on OpenSea. OpenSea broke when we changed `eth_accounts` to throw when unpermitted. We are now preserving legacy `eth_accounts` logic through the legacy method middleware. Read more on the issue here - https://github.com/MetaMask/metamask-mobile/issues/9443
- Go to OpenSea
- Connect an account
- Should experience expected connect flow

To test RPCMethodMiddleware
- Go to MM test dapp
- Trigger any rpc method calls (except eth_accounts)
- App should respond accordingly
- Ex. test wallet_requestPermissions, which is handled within RPCMethodMiddleware

### **After**

In this first video, we're demonstrating the following:
- eth_accounts is hitting the legacy handler via the legacy middleware
- wallet_requestPermissions bypasses both the unsupported method and legacy method middleware and is hitting the RPCMethodMiddleware
- wallet_getPermissions is listed as an unsupported method for testing purposes. We can see that calling it hits the unsupported method middleware
https://github.com/MetaMask/metamask-mobile/assets/10508597/c7e1186b-1626-4724-8aae-203628edf59d

And to show that OpenSea dapp connection still works (did not regress) - https://github.com/MetaMask/metamask-mobile/issues/9443, OpenSea MM connect works as expected

https://github.com/MetaMask/metamask-mobile/assets/10508597/3418f3cf-2b77-48ec-8382-223bac98d15b

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/core/RPCMethods/utils.ts
Comment (Cal-L) [inline app/core/RPCMethods/utils.ts:1]: Tests are captured in `createLegacyMethodMiddleware`, which utilizes these functions

File: app/core/RPCMethods/RPCMethodMiddleware.ts [hunk 1]
```diff
@@ -519,11 +517,6 @@ export const getRpcMethodMiddleware = ({
           },
         });
       },
-      eth_signTransaction: async () => {
```
Comment (Cal-L) [inline app/core/RPCMethods/RPCMethodMiddleware.ts:522]: Now handled by `createUnsupportedMethodMiddleware`

File: app/core/RPCMethods/RPCMethodMiddleware.ts [hunk 2]
```diff
@@ -490,8 +490,6 @@ export const getRpcMethodMiddleware = ({
           }
         }
       },
-      // TODO: This code is no longer used and should be removed in the future. eth_accounts is implemented by permissions specifications
-      eth_accounts: getEthAccounts,
```
Comment (Cal-L) [inline app/core/RPCMethods/RPCMethodMiddleware.ts:494]: Now handled by `createLegacyMethodMiddleware`

File: app/core/RPCMethods/createUnsupportedMethodMiddleware/index.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Creates a middleware that rejects explicitly unsupported RPC methods with the
+ * appropriate error.
+ */
+const createUnsupportedMethodMiddleware = (): JsonRpcMiddleware<
+  unknown,
+  void
+> =>
+  async function unsupportedMethodMiddleware(req, _res, next, end) {
+    if ((UNSUPPORTED_RPC_METHODS as Set<string>).has(req.method)) {
```
Comment (tommasini) [inline app/core/RPCMethods/createUnsupportedMethodMiddleware/index.ts:14]: Would this make it more type safety? For me it's okay to go like this, just wondering 
It also increases the burden of adding new unsupported rpc methods in two places (The set constant and the type here)

```suggestion
    if ((UNSUPPORTED_RPC_METHODS as Set<'eth_signTransaction'>).has(req.method)) {
```

**Review by tommasini — DISMISSED** [review]
This is awesome! LGTM! Just left one nit pick and one thought!

File: app/core/RPCMethods/createUnsupportedMethodMiddleware/index.ts (hunk 3 above)
Comment (rekmarks) [reply app/core/RPCMethods/createUnsupportedMethodMiddleware/index.ts:14]: The problem is that `UNSUPPORTED_RPC_METHODS` is of type `Set<'eth_signTransaction'>` while `req.method` is `string`. The code is good as-is.

File: app/core/RPCMethods/utils.ts [hunk 4]
```diff
@@ -1,5 +1,121 @@
...
+  hooks: Record<string, unknown>,
+  expectedHookNames: Set<string>,
+) {
+  const missingHookNames: string[] = [];
+  expectedHookNames.forEach((hookName) => {
+    if (!hasProperty(hooks, hookName)) {
+      missingHookNames.push(hookName);
+    }
+  });
+  if (missingHookNames.length > 0) {
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply app/core/RPCMethods/utils.ts:31]: We should keep it; it's better to be explicit. `.length > 0` unambiguously communicates that we're looking for non-empty arrays. We should avoid relying on JavaScript's boolean coercion logic when possible.

File: app/core/RPCMethods/utils.ts [hunk 5]
```diff
@@ -1,5 +1,121 @@
...
+    throw new Error(
+      `Missing expected hooks:\n\n${missingHookNames.join('\n')}\n`,
+    );
+  }
+  const extraneousHookNames = Object.getOwnPropertyNames(hooks).filter(
+    (hookName) => !expectedHookNames.has(hookName),
+  );
+  if (extraneousHookNames.length > 0) {
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply app/core/RPCMethods/utils.ts:39]: See above.

File: app/core/RPCMethods/utils.ts [hunk 6]
```diff
@@ -1,5 +1,121 @@
...
+      next,
+      end,
+    ) => {
+      const handler = handlerMap[req.method];
+      if (handler) {
+        const { implementation, hookNames } = handler;
+        try {
+          // Implementations may or may not be async, so we must await them.
+          return await implementation(
+            // @ts-expect-error JsonRpcId (number | string | void) doesn't match PS middleware's id, which is (string | number | null)
```
Comment (rekmarks) [inline app/core/RPCMethods/utils.ts:90]: PS = permission system?

```suggestion
            // @ts-expect-error JsonRpcId (number | string | void) doesn't match the permission middleware's id, which is (string | number | null)
```

**Review by rekmarks — CHANGES_REQUESTED** [review]
This looks good! Just one minor issue: I'm confused by a comment and would like to clarify that first before approving.

**Review by rekmarks — APPROVED** [review]

File: app/core/RPCMethods/utils.ts (hunk 6 above)
Comment (Cal-L) [reply app/core/RPCMethods/utils.ts]: Updated

**Comment by Cal-L** [comment]
Test passed on one of the latest commit with functional changes

**Review by tommasini — APPROVED** [review]

**Comment by christopherferreira9** [comment]
This is currently breaking the SDKs. We're investigating internally what's causing the issue.

**Review by abretonc7s — APPROVED** [review]
LGTM

**Comment by christopherferreira9** [comment]
> This is currently breaking the SDKs. We're investigating internally what's causing the issue.
> 
>  Screen.Recording.2024-06-11.at.11.58.37.mov



> This is currently breaking the SDKs. We're investigating internally what's causing the issue.
> 
>  Screen.Recording.2024-06-11.at.11.58.37.mov

After verifying that the issue was not present on main I've re-checked with a clean build from both the SDk side and wallet and it is working properly. I've sorry for the wrong warning. All good from our side.

**Pull request merged**